modules/acm/src/main/java/edu/gemini/epics/acm/CaTaskControlImpl.java

```diff
@@ -0,0 +1,477 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+public class CaTaskControlImpl implements CaTaskControl {
+    private static final Logger LOG = LoggerFactory.getLogger(CaApplySenderImpl.class
+            .getName());
+
+    private final String name;
+    private final String description;
+
+    private final Boolean trace = true;
+
+    private long timeout;
+    private TimeUnit timeoutUnit;
+    private ScheduledExecutorService executor;
+    private ScheduledFuture<?> timeoutFuture;
+    private final EpicsWriter epicsWriter;
+    private final ChannelListener<Integer> valListener;
+    private final ChannelListener<TaskControlState> stateListener;
+    private final CaParameterStorage storage;
+    private final CaTaskControlRecord record;
+
+    private State currentState = IdleState;
+    private static final State IdleState = new State() {
+        @Override
+        public String signature() {
+            return "IdleState";
+        }
+
+        @Override
+        public State onValChange(Integer val) {
+            return this;
+        }
+
+        @Override
+        public State onStateChange(TaskControlState st) {
+            return this;
+        }
+
+        @Override
+        public State onTimeout() {
+            return this;
+        }
+    };
+
+    public CaTaskControlImpl(String name, String channel, String description, EpicsService epicsService)
+            throws CAException {
+        this.name = name;
+        this.description = description;
+
+        epicsWriter = new EpicsWriterImpl(epicsService);
+        storage = new CaParameterStorage(epicsWriter);
+        record = new CaTaskControlRecord(channel, epicsService);
+
+        record.registerValListener(valListener = new ChannelListener<Integer>() {
+            @Override
+            public void valueChanged(String arg0, List<Integer> newVals) {
+                if (newVals != null && !newVals.isEmpty()) {
+                    CaTaskControlImpl.this.onValChange(newVals.get(0));
+                }
+            }
+        });
+
+        record.registerStateListener(stateListener = new ChannelListener<TaskControlState>() {
+            @Override
+            public void valueChanged(String arg0, List<TaskControlState> newVals) {
+                if (newVals != null && !newVals.isEmpty()) {
+                    CaTaskControlImpl.this.onStateChange(newVals.get(0));
+                }
+            }
+        });
+
+        executor = new ScheduledThreadPoolExecutor(2);
+
+    }
+
+    @Override
+    public String getName() {
+        return name;
+    }
+
+    @Override
+    public String getApply() {
+        return null;
+    }
+
+    @Override
+    public String getCAR() {
+        return null;
```

Comment:
I'm a bit confused here... I thought you said a `taskControl` was a CAD and a CAR. Why does it return `null` for `getCAR()`?

(This could just be me not having done more than a modicum of EPICS programming.)

Reply:
The CAR record is used to publish the execution state of a CAD or Apply record. A taskControl record has an output that tell the execution state of the record. In that way, it absorbs the function of a CAR and does not need t o be associated with one.